documentation/specs/src/interoperability/ethereum-bridge/ethereum_smart_contracts.md

```diff
@@ -1,27 +1,180 @@
+- the total supply of the wNAM ERC20 token, which will represent Namada-native
+  NAM on the EVM chain
+- an initial whitelist of ERC20 tokens that may cross the bridge from Ethereum
+  to Namada - specifically, for each whitelisted ERC20:
+    - the Ethereum address of the ERC20 contract
+    - a cap on the total amount that may cross the bridge, in units of ERC20
+
+After a deployment has finished successfully, the deployer must not have any
+privileged control of any of the contracts deployed. Any privileged actions must
+only be possible via a message signed by a validator set that the smart
+contracts are storing details of.
+
+## Communication
+
+### From Ethereum to Namada
+
+A Namada chain's validators are configured to listen to events emitted by the
+smart contracts pointed to by the _Proxy_ contract. The address of the _Proxy_
+contract is set in a governance parameter in Namada storage. Namada validators
+treat emitted events as authoritative and take action on them. Namada also knows
+the address of the _wNAM_ ERC20 contract via a governance parameter, and treats
+transfers of this ERC20 to Namada as an indication to release native NAM from
+the `#EthBridgeEscrow` account on Namada, rather than to mint a wrapped ERC20 as
+is the case with all other ERC20s.
+
+### From Namada to Ethereum
+
+At any time, the _Governance_ and _Bridge_ contracts must store:
+
+- a hash of the current Namada epoch's active validator set
+- a hash of another epoch's active validator set. When the bridge is first
+  deployed, this will also be the current Namada epoch's active validator set,
+  but after the first valset update is submitted to the _Governance_ smart
+  contract, this hash will always be an adjacent Namada epoch's active validator
+  set i.e. either the previous epoch's, or the next epoch's
+
+In the case of the _Governance_ contract, these are hashes of a map of
+validator's _cold_ key addresses to their voting powers, while for the _Bridge_
+contract it is hashes of a map of validator's _hot_ key addresses to their
+voting powers. Namada validators may post signatures as on chain of relevant
+messages to be relayed to the Ethereum bridge smart contracts (e.g. validator
+set updates, pending transfers, etc.). Methods of the Ethereum bridge smart
+contracts should generally accept:
+
+- some message
+- full details of some active validator set (i.e. relevant Ethereum addresses +
+  voting powers)
+- signatures over the message by validators from the this active validator set
+
+Given this data, anyone should be able to make the relevant Ethereum smart
+contract method call, if they are willing to pay the Ethereum gas. A call is
+then authorized to happen if:
+
+- the active validator set specified in the call hashes to *either* of the
+  validator set hashes stored in the smart contract
+- a quorum (i.e. more than 2/3 by voting power) of the signatures over the
+  message are valid
+
+### Valset updates
```

Comment:
```suggestion
### Validator set updates
```

Reply:
:+1: have changed all usages of valset to validator set